File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/AppMain.java
package com.firefly.example.reactive.coffee.store;

import com.firefly.$;
import com.firefly.annotation.Inject;
import com.firefly.core.ApplicationContext;
import com.firefly.example.reactive.coffee.store.router.RouterInstaller;
import com.firefly.example.reactive.coffee.store.utils.DBUtils;
import com.firefly.server.http2.HTTP2ServerBuilder;
import com.firefly.utils.lang.AbstractLifeCycle;
import com.firefly.utils.log.slf4j.ext.LazyLogger;

/**
 * Firefly coffee store example
 *
 * @author Pengtao Qiu
 */
public class AppMain extends AbstractLifeCycle {

    private static final LazyLogger logger = LazyLogger.create();
    public static final ApplicationContext javaCtx = $.createApplicationContext("java-example.xml");

    @Inject
    private HTTP2ServerBuilder server;

    @Inject
    private ProjectConfig config;

    @Inject
    private DBUtils dbUtils;

    public void initData() {
        dbUtils.createTables();
        dbUtils.initializeData();
    }

    public void installRouters() {
        javaCtx.getBeans(RouterInstaller.class).stream().sorted().forEach(installer -> {
            logger.info(() -> "install routers [" + installer.getClass().getName() + "]");
            installer.install();
        });
    }

    @Override
    protected void init() {
        initData();
        installRouters();
        server.listen(config.getHost(), config.getPort());
    }

    @Override
    protected void destroy() {
        server.stop();
    }

    public static void main(String[] args) {
        javaCtx.getBean(AppMain.class).start();
    }
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/AppMain.kt
package com.firefly.example.kotlin.coffee.store

import com.firefly.annotation.Inject
import com.firefly.example.kotlin.coffee.store.router.RouterInstaller
import com.firefly.example.kotlin.coffee.store.utils.DBUtils
import com.firefly.kotlin.ext.context.Context
import com.firefly.kotlin.ext.context.getBean
import com.firefly.kotlin.ext.http.HttpServer
import com.firefly.kotlin.ext.log.KtLogger
import com.firefly.kotlin.ext.log.info
import com.firefly.utils.lang.AbstractLifeCycle

/**
 * @author Pengtao Qiu
 */

val ktCtx = Context.create("kotlin-example.xml")

class AppMain : AbstractLifeCycle() {

    private val log = KtLogger.getLogger { }

    @Inject
    private lateinit var server: HttpServer

    @Inject
    private lateinit var dbUtils: DBUtils

    @Inject
    private lateinit var projectConfig: ProjectConfig

    override fun init() {
        dbUtils.createTables()
        dbUtils.initializeData()

        ktCtx.getBeans(RouterInstaller::class.java).sorted().forEach {
            log.info { "install routers [${it::class.qualifiedName}]" }
            it.install()
        }
//        server.server.configuration.isSecureConnectionEnabled = true
        server.listen(projectConfig.host, projectConfig.port)
    }

    override fun destroy() {
        server.stop()
    }
}

fun main(args: Array<String>) {
    ktCtx.getBean<AppMain>().start()
}